This failure
```
------- Stdout: -------
=== RUN   TestLogic/3node-tenant/geospatial/geog_operators
--- total progress: 114798 statements/queries
            --- FAIL: TestLogic/3node-tenant/geospatial/geog_operators (0.25s)
                logic.go:2051: 
                     
                    testdata/logic_test/geospatial:2062: SELECT
                      a.dsc,
                      ST_AsEWKT(ST_Centroid(a.geog, false)),
                      ST_AsEWKT(ST_Centroid(a.geog, true))
                    FROM geog_operators_test a
                    ORDER BY a.dsc
                    expected:
                        Empty GeometryCollection                  SRID=4326;GEOMETRYCOLLECTION EMPTY                      SRID=4326;GEOMETRYCOLLECTION EMPTY
                        Empty LineString                          SRID=4326;GEOMETRYCOLLECTION EMPTY                      SRID=4326;GEOMETRYCOLLECTION EMPTY
                        Faraway point                             SRID=4326;POINT (5 5)                                   SRID=4326;POINT (5 5)
                        Line going through left and right square  SRID=4326;POINT (0 0.500019038226216)                   SRID=4326;POINT (0 0.500019038226216)
                        NULL                                      NULL                                                    NULL
                        Point middle of Left Square               SRID=4326;POINT (-0.5 0.5)                              SRID=4326;POINT (-0.5 0.5)
                        Point middle of Right Square              SRID=4326;POINT (0.5 0.5)                               SRID=4326;POINT (0.5 0.5)
                        Square (left)                             SRID=4326;POINT (-0.499999999033419 0.500006343417203)  SRID=4326;POINT (-0.500000083989543 0.500006428376563)
                        Square (right)                            SRID=4326;POINT (0.500000000966581 0.500006343417203)   SRID=4326;POINT (0.499999916010457 0.500006428376563)
                        Square overlapping left and right square  SRID=4326;POINT (0.45000000128654 0.500010341293701)    SRID=4326;POINT (0.449999907835107 0.50001042625301)
                        
                    but found (query options: "") :
                        Empty GeometryCollection                  SRID=4326;GEOMETRYCOLLECTION EMPTY                      SRID=4326;GEOMETRYCOLLECTION EMPTY
                        Empty LineString                          SRID=4326;GEOMETRYCOLLECTION EMPTY                      SRID=4326;GEOMETRYCOLLECTION EMPTY
                        Empty Point                               SRID=4326;GEOMETRYCOLLECTION EMPTY                      SRID=4326;GEOMETRYCOLLECTION EMPTY
                        Faraway point                             SRID=4326;POINT (5 5)                                   SRID=4326;POINT (5 5)
                        Line going through left and right square  SRID=4326;POINT (0 0.500019038226216)                   SRID=4326;POINT (0 0.500019038226216)
                        NULL                                      NULL                                                    NULL
                        Point middle of Left Square               SRID=4326;POINT (-0.5 0.5)                              SRID=4326;POINT (-0.5 0.5)
                        Point middle of Right Square              SRID=4326;POINT (0.5 0.5)                               SRID=4326;POINT (0.5 0.5)
                        Square (left)                             SRID=4326;POINT (-0.499999999033419 0.500006343417203)  SRID=4326;POINT (-0.500000083989543 0.500006428376563)
                        Square (right)                            SRID=4326;POINT (0.500000000966581 0.500006343417203)   SRID=4326;POINT (0.499999916010457 0.500006428376563)
                        Square overlapping left and right square  SRID=4326;POINT (0.45000000128654 0.500010341293701)    SRID=4326;POINT (0.449999907835107 0.50001042625301)
                        
                logic.go:1609: 
                     testdata/logic_test/geospatial:2081: too many errors encountered, skipping the rest of the input
```
seems related to this PR.

bors r-

--- a/docs/generated/sql/functions.md
+++ b/docs/generated/sql/functions.md
@@ -842,6 +842,13 @@ given Geometry.</p>
 <p>This variant approximates the circle into quad_seg segments per line (the default is 8).</p>
 <p>This function utilizes the GEOS module.</p>
 </span></td></tr>
+<tr><td><a name="st_centroid"></a><code>st_centroid(geography: geography) &rarr; geography</code></td><td><span class="funcdesc"><p>Returns the centroid of given geography. Uses a spheroid to perform the operation.</p>
+<p>This function utilizes the GeographicLib library for spheroid calculations.</p>
+</span></td></tr>
+<tr><td><a name="st_centroid"></a><code>st_centroid(geography: geography, use_spheroid: <a href="bool.html">bool</a>) &rarr; geography</code></td><td><span class="funcdesc"><p>Returns the centroid of given geography.</p>
+<p>This function utilizes the S2 library for spherical calculations.</p>
+<p>This function utilizes the GeographicLib library for spheroid calculations.</p>
+</span></td></tr>
 <tr><td><a name="st_centroid"></a><code>st_centroid(geometry: geometry) &rarr; geometry</code></td><td><span class="funcdesc"><p>Returns the centroid of the given geometry.</p>
 <p>This function utilizes the GEOS module.</p>
 </span></td></tr>

--- a/pkg/geo/geogfn/topology_operations.go
+++ b/pkg/geo/geogfn/topology_operations.go
@@ -1,0 +1,122 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package geogfn
+
+import (
+	"github.com/cockroachdb/cockroach/pkg/geo"
+	"github.com/cockroachdb/errors"
+	"github.com/golang/geo/r3"
+	"github.com/golang/geo/s2"
+	"github.com/twpayne/go-geom"
+)
+
+// Centroid returns the Centroid of a given Geography.
+//
+// NOTE: In the case of (Multi)Polygon Centroid result, it doesn't mirror with
+// PostGIS's result. We are using the same algorithm of dividing into triangles.
+// However, The PostGIS implementation differs as it cuts triangles in a different
+// way - namely, it fixes the first point in the exterior ring as the first point
+// of the triangle, whereas we always update the reference point to be
+// the first point of the ring when moving from one ring to another.
+//
+// See: http://jennessent.com/downloads/Graphics_Shapes_Manual_A4.pdf#page=49
+// for more details.
+//
+// Ideally, both implementations should provide the same result. However, the
+// centroid of the triangles is the vectorized mean of all the points, not the
+// actual projection in the Spherical surface, which causes a small inaccuracies.
+// This inaccuracy will eventually grow if there is a substantial
+// number of a triangle with a larger area.
+func Centroid(g *geo.Geography, useSphereOrSpheroid UseSphereOrSpheroid) (*geo.Geography, error) {
+	geomRepr, err := g.AsGeomT()
+	if err != nil {
+		return nil, err
+	}
+	if geomRepr.Empty() {
+		return geo.NewGeographyFromGeom(geom.NewGeometryCollection().SetSRID(geomRepr.SRID()))
+	}
+	switch geomRepr.(type) {
+	case *geom.Point, *geom.LineString, *geom.Polygon, *geom.MultiPoint, *geom.MultiLineString, *geom.MultiPolygon:
+	default:
+		return nil, errors.Newf("unhandled geography type %s", g.Shape().String())
+	}
+
+	regions, err := geo.S2RegionsFromGeom(geomRepr, geo.EmptyBehaviorOmit)
+	if err != nil {
+		return nil, err
+	}
+	spheroid, err := g.Spheroid()
+	if err != nil {
+		return nil, err
+	}
+
+	// localWeightedCentroids is the collection of all the centroid corresponds to
+	// various small regions in which we divide the given region for calculation
+	// of centroid. The magnitude of each s2.Point.Vector represents
+	// the weight corresponding to its region.
+	var localWeightedCentroids []s2.Point
+	for _, region := range regions {
+		switch region := region.(type) {
+		case s2.Point:
+			localWeightedCentroids = append(localWeightedCentroids, region)
+		case *s2.Polyline:
+			// The algorithm used for the calculation of centroid for (Multi)LineString:
+			//  * Split (Multi)LineString in the set of individual edges.
+			//  * Calculate the mid-points and length/angle for all the edges.
+			//  * The centroid of (Multi)LineString will be a weighted average of mid-points
+			//    of all the edges, where each mid-points is weighted by its length/angle.
+			for edgeIdx := 0; edgeIdx < region.NumEdges(); edgeIdx++ {
+				var edgeWeight float64
+				eV0 := region.Edge(edgeIdx).V0
+				eV1 := region.Edge(edgeIdx).V1
+				if useSphereOrSpheroid == UseSpheroid {
+					edgeWeight = spheroidDistance(spheroid, eV0, eV1)
+				} else {
+					edgeWeight = float64(s2.ChordAngleBetweenPoints(eV0, eV1).Angle())
+				}
+				localWeightedCentroids = append(localWeightedCentroids, s2.Point{Vector: eV0.Add(eV1.Vector).Mul(edgeWeight)})
+			}
+		case *s2.Polygon:
+			// The algorithm used for the calculation of centroid for (Multi)Polygon:
+			//  * Split (Multi)Polygon in the set of individual triangles.
+			//  * Calculate the centroid and signed area (negative area for triangle inside
+			//    the hole) for all the triangle.
+			//  * The centroid of (Multi)Polygon will be a weighted average of the centroid
+			//    of all the triangle, where each centroid is weighted by its area.
+			for _, loop := range region.Loops() {
+				triangleVertices := make([]s2.Point, 4)
+				triangleVertices[0] = loop.Vertex(0)
+				triangleVertices[3] = loop.Vertex(0)
+
+				for pointIdx := 1; pointIdx+2 < loop.NumVertices(); pointIdx++ {
+					triangleVertices[1] = loop.Vertex(pointIdx)
+					triangleVertices[2] = loop.Vertex(pointIdx + 1)
+					triangleCentroid := s2.PlanarCentroid(triangleVertices[0], triangleVertices[1], triangleVertices[2])
+					var area float64
+					if useSphereOrSpheroid == UseSpheroid {
+						area, _ = spheroid.AreaAndPerimeter(triangleVertices[:3])
+					} else {
+						area = s2.LoopFromPoints(triangleVertices).Area()
+					}
+					area = area * float64(loop.Sign())
+					localWeightedCentroids = append(localWeightedCentroids, s2.Point{Vector: triangleCentroid.Mul(area)})
+				}
+			}
+		}
+	}
+	var centroidVector r3.Vector
+	for _, point := range localWeightedCentroids {
+		centroidVector = centroidVector.Add(point.Vector)
+	}
+	latLng := s2.LatLngFromPoint(s2.Point{Vector: centroidVector.Normalize()})
+	centroid := geom.NewPointFlat(geom.XY, []float64{latLng.Lng.Degrees(), latLng.Lat.Degrees()}).SetSRID(int(g.SRID()))
+	return geo.NewGeographyFromGeom(centroid)
+}

--- a/pkg/geo/geogfn/topology_operations_test.go
+++ b/pkg/geo/geogfn/topology_operations_test.go
@@ -1,0 +1,160 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package geogfn
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/geo"
+	"github.com/stretchr/testify/require"
+	"github.com/twpayne/go-geom"
+)
+
+func TestCentroid(t *testing.T) {
+	testCases := []struct {
+		wkt                      string
+		expectedSphereCentroid   string
+		expectedSpheroidCentroid string
+	}{
+		{
+			wkt:                      `POINT (1.0 1.0)`,
+			expectedSphereCentroid:   `POINT (1.0 1.0)`,
+			expectedSpheroidCentroid: `POINT (1.0 1.0)`,
+		},
+		{
+			wkt:                      `SRID=4326;POINT (1.0 1.0)`,
+			expectedSphereCentroid:   `SRID=4326;POINT (1.0 1.0)`,
+			expectedSpheroidCentroid: `SRID=4326;POINT (1.0 1.0)`,
+		},
+		{
+			wkt:                      `LINESTRING (1.0 1.0, 2.0 2.0, 3.0 3.0)`,
+			expectedSphereCentroid:   `POINT (1.99961910254054 2.00007602871312)`,
+			expectedSpheroidCentroid: `POINT (1.99962062254316 2.00007754894762)`,
+		},
+		{
+			wkt:                      `LINESTRING (-1.0 -1.0, 1.0 -1.0, 1.0 -30.0, -1.0 -30.0)`,
+			expectedSphereCentroid:   `POINT(0.885490683311801 -15.3793090065665)`,
+			expectedSpheroidCentroid: `POINT(0.884862829191911 -15.3793503404433)`,
+		},
+		{
+			wkt:                      `POLYGON ((0.0 0.0, 1.0 0.0, 1.0 1.0, 0.0 1.0, 0.0 0.0))`,
+			expectedSphereCentroid:   `POINT (0.499999916440062 0.500006427946943)`,
+			expectedSpheroidCentroid: `POINT (0.500000001396299 0.500006342987471)`,
+		},
+		{
+			wkt:                      `POLYGON ((1.0 1.0, 0.0 1.0, 0.0 0.0, 1.0 0.0, 1.0 1.0))`,
+			expectedSphereCentroid:   `POINT (0.499999916440062 0.500006427946943)`,
+			expectedSpheroidCentroid: `POINT (0.500000001396299 0.500006342987471)`,
+		},
+		{
+			wkt:                      `POLYGON ((0.0 0.0, 80.0 0.0, 80.0 80.0, 0.0 0.0))`,
+			expectedSphereCentroid:   `POINT (43.8350223319301 30.5453774409703)`,
+			expectedSpheroidCentroid: `POINT (43.8350223319301 30.5453774409703)`,
+		},
+		{
+			wkt:                      `POLYGON ((0.0 0.0, 1.0 0.0, 1.0 1.0, 0.0 1.0, 0.0 0.0), (0.1 0.1, 0.9 0.1, 0.9 0.9, 0.1 0.9, 0.1 0.1))`,
+			expectedSphereCentroid:   `POINT (0.500000001981181 0.500010402851408)`,
+			expectedSpheroidCentroid: `POINT (0.499999862652707 0.500010542184438)`,
+		},
+		{
+			wkt:                      `POLYGON ((1.0 1.0, 0.0 1.0, 0.0 0.0, 1.0 0.0, 1.0 1.0), (0.1 0.1, 0.9 0.1, 0.9 0.9, 0.1 0.9, 0.1 0.1))`,
+			expectedSphereCentroid:   `POINT (0.500000001981181 0.500010402851408)`,
+			expectedSpheroidCentroid: `POINT (0.499999862652707 0.500010542184438)`,
+		},
+		{
+			wkt:                      `POLYGON ((0.0 1.0, 0.0 0.0, 1.0 0.0, 1.0 1.0, 0.0 1.0), (0.9 0.9, 0.1 0.9, 0.1 0.1, 0.9 0.1, 0.9 0.9))`,
+			expectedSphereCentroid:   `POINT (0.500000001981181 0.500010402851408)`,
+			expectedSpheroidCentroid: `POINT (0.499999862652707 0.500010542184438)`,
+		},
+		{
+			wkt:                      `POLYGON ((0.0 0.0, -3.0 0.0, -4.0 -4.0, 0.0 -3.0, 0.0 0.0), (-1.0 -1.0, -2.0 -1.0, -2.0 -2.0, -1.0 -2.0, -1.0 -1.0))`,
+			expectedSphereCentroid:   `POINT (-1.86284249441196 -1.864020960242764)`,
+			expectedSpheroidCentroid: `POINT (-1.862840114700246 -1.864022599497527)`,
+		},
+		{
+			wkt:                      `MULTIPOINT (0.0 0.0, 1.0 10.0, 2.0 30.0)`,
+			expectedSphereCentroid:   `POINT (0.953002819213836 13.2963208764279)`,
+			expectedSpheroidCentroid: `POINT (0.953002819213836 13.2963208764279)`,
+		},
+		{
+			wkt:                      `MULTILINESTRING ((1.0 1.0, 2.0 2.0, 3.0 3.0), (0.0 0.0, 1.0 10.0, 2.0 30.0))`,
+			expectedSphereCentroid:   `POINT (1.20651354391928 13.8361533064176)`,
+			expectedSpheroidCentroid: `POINT (1.20691326039874 13.8369776044908)`,
+		},
+		{
+			wkt:                      `MULTIPOLYGON (((0.0 0.0, 1.0 0.0, 1.0 1.0, 0.0 1.0, 0.0 0.0)), ((10.0 10.0, 14.0 10.0, 14.0 14.0, 10.0 14.0, 10.0 10.0)))`,
+			expectedSphereCentroid:   `POINT (11.298142253801448 11.327334347758455)`,
+			expectedSpheroidCentroid: `POINT (11.298497717682135 11.327735897664535)`,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.wkt, func(t *testing.T) {
+			geog, err := geo.ParseGeography(tc.wkt)
+			require.NoError(t, err)
+
+			retSphereCentroid, err := Centroid(geog, UseSphere)
+			require.NoError(t, err)
+			retGeomSphereCentroid, err := retSphereCentroid.AsGeomT()
+			require.NoError(t, err)
+			retSpheroidCentroid, err := Centroid(geog, UseSpheroid)
+			require.NoError(t, err)
+			retGeomSpheroidCentroid, err := retSpheroidCentroid.AsGeomT()
+			require.NoError(t, err)
+
+			expSphereCentroid, err := geo.ParseGeography(tc.expectedSphereCentroid)
+			require.NoError(t, err)
+			expGeomSphereCentroid, err := expSphereCentroid.AsGeomT()
+			require.NoError(t, err)
+			expSpheroidCentroid, err := geo.ParseGeography(tc.expectedSpheroidCentroid)
+			require.NoError(t, err)
+			expGeomSpheroidCentroid, err := expSpheroidCentroid.AsGeomT()
+			require.NoError(t, err)
+
+			// Ensure points are close in terms of precision.
+			require.InEpsilon(t, expGeomSphereCentroid.(*geom.Point).X(), retGeomSphereCentroid.(*geom.Point).X(), 1e-6)
+			require.InEpsilon(t, expGeomSphereCentroid.(*geom.Point).Y(), retGeomSphereCentroid.(*geom.Point).Y(), 1e-6)
+			require.Equal(t, retSpheroidCentroid.SRID(), expSphereCentroid.SRID())
+			require.InEpsilon(t, expGeomSpheroidCentroid.(*geom.Point).X(), retGeomSpheroidCentroid.(*geom.Point).X(), 1e-6)
+			require.InEpsilon(t, expGeomSpheroidCentroid.(*geom.Point).Y(), retGeomSpheroidCentroid.(*geom.Point).Y(), 1e-6)
+			require.Equal(t, retSpheroidCentroid.SRID(), expSphereCentroid.SRID())
+		})
+	}
+
+	t.Run("EMPTY GeometryCollection for EMPTY Geographical object", func(t *testing.T) {
+		for _, wkt := range []string{
+			`LINESTRING EMPTY`,
+			`POLYGON EMPTY`,
+			`GEOMETRYCOLLECTION EMPTY`,
+			`GEOMETRYCOLLECTION (LINESTRING EMPTY, POLYGON EMPTY)`,
+		} {
+			for _, useSphereOrSpheroid := range []UseSphereOrSpheroid{UseSphere, UseSpheroid} {
+				t.Run(fmt.Sprintf("Centroid(geography=%s, spheroid=%t)", wkt, useSphereOrSpheroid), func(t *testing.T) {
+					geog, err := geo.ParseGeography(wkt)
+					require.NoError(t, err)
+					ret, err := Centroid(geog, useSphereOrSpheroid)
+					require.NoError(t, err)
+					exp, err := geo.ParseGeography(`GEOMETRYCOLLECTION EMPTY`)
+					require.NoError(t, err)
+					require.Equal(t, ret, exp)
+				})
+			}
+		}
+	})
+
+	t.Run(`error for Non EMPTY GeometryCollection`, func(t *testing.T) {
+		geog, err := geo.ParseGeography(`GEOMETRYCOLLECTION (POINT (0.0 0.0), LINESTRING EMPTY)`)
+		require.NoError(t, err)
+		_, err = Centroid(geog, UseSpheroid)
+		require.EqualError(t, err, "unhandled geography type GeometryCollection")
+	})
+}

--- a/pkg/sql/logictest/testdata/logic_test/geospatial
+++ b/pkg/sql/logictest/testdata/logic_test/geospatial
@@ -2057,6 +2057,30 @@ Square (left)                             POLYGON ((-1 0, -0.5 0, 0 0, 0 0.5, 0 
 Square (right)                            POLYGON ((0 0, 0.5 0, 1 0, 1 0.5, 1 1, 0.5 1.000038070652873, 0 1, 0 0.5, 0 0))                POLYGON ((0 0, 0.25 0, 0.5 0, 0.75 0, 1 0, 1 0.25, 1 0.5, 1 0.75, 1 1, 0.750000001449807 1.000028552944327, 0.5 1.000038070652873, 0.249999998550193 1.000028552944327, 0 1, 0 0.75, 0 0.5, 0 0.25, 0 0))
 Square overlapping left and right square  POLYGON ((-0.1 0, 0.45 0, 1 0, 1 0.5, 1 1, 0.45 1.000046065796834, -0.1 1, -0.1 0.5, -0.1 0))  POLYGON ((-0.1 0, 0.175 0, 0.45 0, 0.725 0, 1 0, 1 0.25, 1 0.5, 1 0.75, 1 1, 0.725000001929716 1.000034549281259, 0.45 1.000046065796834, 0.174999998070284 1.000034549281259, -0.1 1, -0.1 0.75, -0.1 0.5, -0.1 0.25, -0.1 0))
 
+# ST_Centroid for Geography
+query TTT
+SELECT
+  a.dsc,
+  ST_AsEWKT(ST_Centroid(a.geog, false)),
+  ST_AsEWKT(ST_Centroid(a.geog, true))
+FROM geog_operators_test a
+ORDER BY a.dsc
+----
+Empty GeometryCollection                  SRID=4326;GEOMETRYCOLLECTION EMPTY                      SRID=4326;GEOMETRYCOLLECTION EMPTY
+Empty LineString                          SRID=4326;GEOMETRYCOLLECTION EMPTY                      SRID=4326;GEOMETRYCOLLECTION EMPTY
+Empty Point                               SRID=4326;GEOMETRYCOLLECTION EMPTY                      SRID=4326;GEOMETRYCOLLECTION EMPTY
+Faraway point                             SRID=4326;POINT (5 5)                                   SRID=4326;POINT (5 5)
+Line going through left and right square  SRID=4326;POINT (0 0.500019038226216)                   SRID=4326;POINT (0 0.500019038226216)
+NULL                                      NULL                                                    NULL
+Point middle of Left Square               SRID=4326;POINT (-0.5 0.5)                              SRID=4326;POINT (-0.5 0.5)
+Point middle of Right Square              SRID=4326;POINT (0.5 0.5)                               SRID=4326;POINT (0.5 0.5)
+Square (left)                             SRID=4326;POINT (-0.499999999033419 0.500006343417203)  SRID=4326;POINT (-0.500000083989543 0.500006428376563)
+Square (right)                            SRID=4326;POINT (0.500000000966581 0.500006343417203)   SRID=4326;POINT (0.499999916010457 0.500006428376563)
+Square overlapping left and right square  SRID=4326;POINT (0.45000000128654 0.500010341293701)    SRID=4326;POINT (0.449999907835107 0.50001042625301)
+
+statement error st_centroid\(\): unhandled geography type GeometryCollection
+SELECT ST_Centroid('GEOMETRYCOLLECTION(POINT(0 0), LINESTRING EMPTY)'::geography, true)
+
 query T
 SELECT ST_AsText(ST_Segmentize('MULTIPOINT (0 0, 1 1)'::geography, -1))
 ----

--- a/pkg/sql/sem/builtins/geo_builtins.go
+++ b/pkg/sql/sem/builtins/geo_builtins.go
@@ -2208,40 +2208,56 @@ Note if geometries are the same, it will return the LineString with the minimum 
 	// Topology operations
 	"st_centroid": makeBuiltin(
 		defProps(),
-		geometryOverload1(
-			func(ctx *tree.EvalContext, g *tree.DGeometry) (tree.Datum, error) {
-				centroid, err := geomfn.Centroid(g.Geometry)
-				if err != nil {
-					return nil, err
-				}
-				return tree.NewDGeometry(centroid), err
-			},
-			types.Geometry,
-			infoBuilder{
-				info:         "Returns the centroid of the given geometry.",
-				libraryUsage: usesGEOS,
-			},
-			tree.VolatilityImmutable,
-		),
-		stringOverload1(
-			func(ctx *tree.EvalContext, s string) (tree.Datum, error) {
-				g, err := geo.ParseGeometry(s)
-				if err != nil {
-					return nil, err
-				}
-				centroid, err := geomfn.Centroid(g)
-				if err != nil {
-					return nil, err
-				}
-				return tree.NewDGeometry(centroid), err
-			},
-			types.Geometry,
-			infoBuilder{
-				info:         "Returns the centroid of the given string, which will be parsed as a geometry object.",
-				libraryUsage: usesGEOS,
-			}.String(),
-			tree.VolatilityImmutable,
-		),
+		append(
+			geographyOverload1WithUseSpheroid(
+				func(ctx *tree.EvalContext, g *tree.DGeography, useSphereOrSpheroid geogfn.UseSphereOrSpheroid) (tree.Datum, error) {
+					ret, err := geogfn.Centroid(g.Geography, useSphereOrSpheroid)
+					if err != nil {
+						return nil, err
+					}
+					return tree.NewDGeography(ret), nil
+				},
+				types.Geography,
+				infoBuilder{
+					info: "Returns the centroid of given geography.",
+				},
+				tree.VolatilityImmutable,
+			),
+			geometryOverload1(
+				func(ctx *tree.EvalContext, g *tree.DGeometry) (tree.Datum, error) {
+					centroid, err := geomfn.Centroid(g.Geometry)
+					if err != nil {
+						return nil, err
+					}
+					return tree.NewDGeometry(centroid), err
+				},
+				types.Geometry,
+				infoBuilder{
+					info:         "Returns the centroid of the given geometry.",
+					libraryUsage: usesGEOS,
+				},
+				tree.VolatilityImmutable,
+			),
+			stringOverload1(
+				func(ctx *tree.EvalContext, s string) (tree.Datum, error) {
+					g, err := geo.ParseGeometry(s)
+					if err != nil {
+						return nil, err
+					}
+					centroid, err := geomfn.Centroid(g)
+					if err != nil {
+						return nil, err
+					}
+					return tree.NewDGeometry(centroid), err
+				},
+				types.Geometry,
+				infoBuilder{
+					info:         "Returns the centroid of the given string, which will be parsed as a geometry object.",
+					libraryUsage: usesGEOS,
+				}.String(),
+				tree.VolatilityImmutable,
+			),
+		)...,
 	),
 	"st_pointonsurface": makeBuiltin(
 		defProps(),

